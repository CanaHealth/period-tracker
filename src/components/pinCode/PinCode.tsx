import React from 'react';
import { useState } from 'react';

import clsxm from '@/lib/clsxm';

import PinInput from '@/components/pinCode/PinInput';

type PinCodeProps = {
  pincode: number[];
  className?: string;
} & React.ComponentPropsWithoutRef<'div'>;

const PinCode: React.FC<PinCodeProps> = ({ className, pincode, ...rest }) => {
  // input handle functions ...

  const [pin, setPin] = useState<number[]>(pincode);
  const [refIndex, setRefIndex] = useState<number>(0);

  const onChangeDigits = (value: number, index: number) => {
    setPin(pin.map((digit, i) => (i === index ? value : digit)));
  };

  const onFocus = (index: number) => {
    setRefIndex(index);
  };

  const onBlur = () => {
    setRefIndex(0);
  };

  // use onKeyDown to check if the user pressed left or right arrow keys to change the focus to the next or previous instance of PinInput component.
  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'ArrowLeft') {
      event.preventDefault();
      setRefIndex(refIndex - 1);
    } else if (event.key === 'ArrowRight') {
      event.preventDefault();
      setRefIndex(refIndex + 1);
    }
  };

  return (
    <div className={clsxm('', className)} {...rest} onKeyDown={handleKeyDown}>
      <div className='flex min-w-full flex-row justify-between'>
        {pin.map((digit, index) => (
          <PinInput
            key={index}
            myRef={index === refIndex}
            placeholder='0'
            onChangeDigit={(digit) => onChangeDigits(digit, index)}
            onFocus={() => onFocus(index)}
            onBlur={onBlur}
            className=''
          />
        ))}
      </div>
    </div>
  );
};

export default PinCode;
