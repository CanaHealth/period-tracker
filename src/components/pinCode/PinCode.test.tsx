import { render } from '@testing-library/react';
import React from 'react';

import PinCode from '@/components/pinCode/PinCode';

describe('PinCode', () => {
  it('should render', () => {
    const { container } = render(
      <PinCode className='' pincode={[0, 0, 0, 0, 0, 0]} />
    );
    expect(container).toMatchSnapshot();
  });
}); //render the PinCode component with a pincode of [0, 0, 0, 0, 0, 0]
