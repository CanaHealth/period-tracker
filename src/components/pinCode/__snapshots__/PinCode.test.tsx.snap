// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinCode should render 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="flex min-w-full flex-row justify-between"
    >
      <input
        class="h-24 w-24 appearance-none rounded border-2 border-gray-300 py-2 px-4 text-center text-5xl leading-tight text-gray-700 shadow-sm focus:border-blue-500 focus:outline-none"
        placeholder="0"
        type="number"
        value="0"
      />
      <input
        class="h-24 w-24 appearance-none rounded border-2 border-gray-300 py-2 px-4 text-center text-5xl leading-tight text-gray-700 shadow-sm focus:border-blue-500 focus:outline-none"
        placeholder="0"
        type="number"
        value="0"
      />
      <input
        class="h-24 w-24 appearance-none rounded border-2 border-gray-300 py-2 px-4 text-center text-5xl leading-tight text-gray-700 shadow-sm focus:border-blue-500 focus:outline-none"
        placeholder="0"
        type="number"
        value="0"
      />
      <input
        class="h-24 w-24 appearance-none rounded border-2 border-gray-300 py-2 px-4 text-center text-5xl leading-tight text-gray-700 shadow-sm focus:border-blue-500 focus:outline-none"
        placeholder="0"
        type="number"
        value="0"
      />
      <input
        class="h-24 w-24 appearance-none rounded border-2 border-gray-300 py-2 px-4 text-center text-5xl leading-tight text-gray-700 shadow-sm focus:border-blue-500 focus:outline-none"
        placeholder="0"
        type="number"
        value="0"
      />
      <input
        class="h-24 w-24 appearance-none rounded border-2 border-gray-300 py-2 px-4 text-center text-5xl leading-tight text-gray-700 shadow-sm focus:border-blue-500 focus:outline-none"
        placeholder="0"
        type="number"
        value="0"
      />
    </div>
  </div>
</div>
`;
